src/dds_service.py

```diff
@@ -72,3 +74,315 @@ def builtin_observer(domain_id, dds_data, running):
+    return None
+
+def ddsi_patmatch(pat, s):
+    pat_idx = 0
+    s_idx = 0
+
+    while pat_idx < len(pat):
+        if pat[pat_idx] == '?':
+            # any character will do
+            if s_idx >= len(s):
+                return False
+            pat_idx += 1
+            s_idx += 1
+        elif pat[pat_idx] == '*':
+            # collapse a sequence of wildcards
+            while pat_idx < len(pat) and (pat[pat_idx] == '*' or pat[pat_idx] == '?'):
+                if pat[pat_idx] == '?' and s_idx >= len(s):
+                    return False
+                pat_idx += 1
+                s_idx += 1
+            # try matching on all positions where s matches pat
+            while s_idx < len(s):
+                if (pat_idx < len(pat) and s[s_idx] == pat[pat_idx] and
+                        ddsi_patmatch(pat[pat_idx + 1:], s[s_idx + 1:])):
+                    return True
+                s_idx += 1
+            return pat_idx == len(pat)
+        else:
+            # only an exact match
+            if s_idx >= len(s) or s[s_idx] != pat[pat_idx]:
+                return False
+            pat_idx += 1
+            s_idx += 1
+
+    return s_idx == len(s)
+
+def is_wildcard_partition(s):
+    return '*' in s or '?' in s
+
+def partition_patmatch_p(pat, name):
+    if not is_wildcard_partition(pat):
+        return pat == name
+    elif is_wildcard_partition(name):
+        return False
+    else:
+        return ddsi_patmatch(pat, name)
+
+def partitions_match_default(x):
+    if qos.Policy.Partition not in x or len(x[qos.Policy.Partition].partitions) == 0:
+        return True
+    for i in range(len(x.partition)):
+        if partition_patmatch_p(x[qos.Policy.Partition].partitions[i], ""):
+            return True
+    return False
+
+def partitions_match_p(a, b):
+    if (qos.Policy.Partition not in a) or len(a[qos.Policy.Partition].partitions) == 0:
+        return partitions_match_default(b)
+    elif (qos.Policy.Partition not in b) or len(b[qos.Policy.Partition].partitions) == 0:
+        return partitions_match_default(a)
+    else:
+        for i in range(len(a[qos.Policy.Partition].partitions)):
+            for j in range(len(b[qos.Policy.Partition].partitions)):
+                if (partition_patmatch_p(a[qos.Policy.Partition].partitions[i], b[qos.Policy.Partition].partitions[j]) or 
+                        partition_patmatch_p(b[qos.Policy.Partition].partitions[j], a[qos.Policy.Partition].partitions[i])):
+                    return True
+        return False
+
+def qos_match(endpoint_reader, endpoint_writer):
+
+    if endpoint_reader.topic_name != endpoint_writer.topic_name:
+        return False, dds_qos_policy_id.DDS_INVALID_QOS_POLICY_ID
+
+    reliability_rd = to_kind_reliability(endpoint_reader.qos)
+    reliability_wr = to_kind_reliability(endpoint_writer.qos)
+    if reliability_rd and reliability_wr and reliability_rd > reliability_wr:
```

Comment:
I think it would be nicer if this returned a list of mismatching QoS (perhaps it would then no longer be called `qos_match`, but that's a detail): usually there'll only be a single QoS causing the mismatch but it could potentially be multiple.

I also wonder if it wouldn't be more practical to return the actual mismatching QoS objects instead of a policy id. We have these sparse QoS objects, so why not return a pair of QoS objects with just the mismatching ones filled in? Then the set of policies in those two objects would always be the same and converting to text would be a simple iteration over one and a conversion to text for both.

There's quite a decent chance that I overlooked something. Perhaps a show stopper, or perhaps it turns out to be more complex in practice.

Reply:
thats true, currently it just returns a single mismatch reason, i will change it so that all mismatches can be seen at once.

Reply:
mhm it would not work for all cases, its not always a qos object, sometimes its e.g. topic_name or topic_type..
I think i keep returning the policy id in a list.
In the gui i already have the qos object which can i show when its marked wit a mismatching policy id.